### DISCUSSION TITLE
Avoid handle `BaseException` just for cover async task cancellation.

### DISCUSSION THREAD

**Discussion by T-256** [encode/httpcore#805] [Potential Issue] [answered]
Actually, I noticed by https://github.com/encode/httpcore/pull/802/files#r1326095576.

In https://github.com/encode/httpcore/pull/802, he tries to handle all of `BaseExecption` errors that we don't handle currently (e.g. `KeyboardInterrupt`).
What's need here for error handling is just async-backend specified cancellation error class.

Actually, not only here, also where https://github.com/encode/httpcore/pull/726 implemented:
- [First](https://github.com/encode/httpcore/pull/726/files#diff-4a314fb3c706f5b1f148fe26dffa6ef7ab62eaaa4217b90ff5fd286b73d0087bR106), _Before_: no error handling, _After_: `BaseException` handling.
- [Second](https://github.com/encode/httpcore/pull/726/files#diff-4a314fb3c706f5b1f148fe26dffa6ef7ab62eaaa4217b90ff5fd286b73d0087bR163), _Before_: `Exception` handling, _After_: `BaseException` handling.

**Comment by lovelydinosaur** [comment]
Can you describe and demonstrate a behaviour that you want to see changed?

**Reply by T-256** [reply]
For example [here](https://github.com/encode/httpcore/pull/726/files#diff-4a314fb3c706f5b1f148fe26dffa6ef7ab62eaaa4217b90ff5fd286b73d0087bR163), you was trying to add support anyio/trio cancel exception, but eventually you covered all other `BaseException` errors.

I think the correct way here is using `Exception` class + `async cancellation` class.

Perhaps we could use [anyio.get_cancelled_exc_class](https://anyio.readthedocs.io/en/stable/cancellation.html#:~:text=In%20some%20specific,~get_cancelled_exc_class%3A) here:
```diff
-       except BaseException as exc:  # noqa: PIE786
+       except (Exception, self._cancelled_class) as exc:  # noqa: PIE786
            with AsyncShieldCancellation():
                kwargs = {"stream_id": stream_id}
                async with Trace("response_closed", logger, request, kwargs):
                    await self._response_closed(stream_id=stream_id)
```
where `self._cancelled_class` defined on constructing the class by using `sniffio`, like what happened inside `AsyncShieldCancellation`.

**Reply by zanieb** [reply]
You're still focused on the implementation here. What user facing behavior is incorrect in the given implementation?

**Reply by T-256** [reply]
From https://docs.python.org/3/library/exceptions.html:

> programmers are encouraged to derive new exceptions from the [Exception](https://docs.python.org/3/library/exceptions.html#Exception) class or one of its subclasses, and not from [BaseException](https://docs.python.org/3/library/exceptions.html#BaseException).

As said, we should never catch `BaseException`.
Even we know trying code raises subclass of `BaseException`, we should catch that class directly.

> What user facing behavior is incorrect in the given implementation?

`GeneratorExit`, `KeyboardInterrupt`, `SystemExit` are builtin exceptions that directly inherited from `BaseException`, can you point why catching such exceptions is good in httpcore?
Or can you say the reason my proposed change is worse than current implementation?

**Reply by T-256** [reply]
Also in [pep-0352](https://peps.python.org/pep-0352/) clarified why `BaseException` exists.

**Reply by lovelydinosaur** [reply]
These are properly scoped catch-cleanup-reraise cases.

**Reply by T-256** [reply]
Let me tell explicitly, `BaseException` literal is standing to be subclassed, not exception handling!

> These are properly scoped catch-cleanup-reraise cases.

Almost entire codebase is using `Exception` for catch-cleanup-reraise (why not `BaseException`?):
- https://github.com/encode/httpcore/blob/a37d7a2ec6c846413eca6c0b492b0384d1ef0877/httpcore/_sync/http2.py#L441
- https://github.com/encode/httpcore/blob/a37d7a2ec6c846413eca6c0b492b0384d1ef0877/httpcore/_sync/http2.py#L470
- https://github.com/encode/httpcore/blob/a37d7a2ec6c846413eca6c0b492b0384d1ef0877/httpcore/_backends/anyio.py#L76
- etc.

And these are correct exception handling!

**Reply by T-256** [reply]
Also as said in https://stackoverflow.com/a/44655928, `except BaseException:` has same functionality as bare `except:`. linters avoid you to use bare `except:` ([1](https://docs.astral.sh/ruff/rules/bare-except/), [2](https://docs.astral.sh/ruff/rules/blind-except/)), so perhaps you now understand what actually I'm pointing to.

**Reply by karpetrosyan** [reply]
> GeneratorExit, KeyboardInterrupt, SystemExit are builtin exceptions that directly inherited from BaseException, can you point why catching such exceptions is good in httpcore?

You can even use bare `except:` when writing your cleanup logic. It is always preferable to clean up even if a keyboardinterrupt has been raised, because the user may catch the keyboardinterrupt in his code and continue to use httpcore.ConnectionPool.

**Reply by T-256** [reply]
> It is always preferable to clean up even if a keyboardinterrupt has been raised,

It's reasonable, but for that purpose, IMO you could use direct class name.

> You can even use bare `except:` when writing your cleanup logic.

but for cleanup purposes you should be able to do cleanup from most top call layer, for an example please check:
https://github.com/T-256/httpcore/tree/cleanup-redesign

**Reply by lovelydinosaur** [reply]
Okay, so I'm not *against* a proposal to switch cases of...

```python
except BaseException as exc:
```

...to...

```python
except (Exception, anyio.get_cancelled_exc_class()) as exc:
```

...what I *would* like to see is a proposal that *demonstrates* that change in a minimal way.

The `cleanup-redesign` branch you've linked above has a commit [across 7 files with 272 additions and 236 deletions](https://github.com/T-256/httpcore/commit/6105e0cd0edb19cfdb6526e5f7109a1d0ceea1c4). That's not proportional to the change you're suggesting.

You're suggesting a potentially valid but not *obviously necessary* change here. That's okay, but if you could present it in a way that helps make it clear you've thought this through thoroughly, and that doesn't attention-churn the rest of the team, that would be much appreciated. 🙂

**Reply by T-256** [reply]
> The `cleanup-redesign` branch you've linked above has a commit [across 7 files with 272 additions and 236 deletions](https://github.com/T-256/httpcore/commit/6105e0cd0edb19cfdb6526e5f7109a1d0ceea1c4). That's not proportional to the change you're suggesting.

Correct, actually that branch is suggesting new cleanup structure, it exists to answer @karosis88 that we can omit `except:` for cleanup and use `_cleanup` method from parent caller.

> if you could present it in a way that helps make it clear you've thought this through thoroughly, and that doesn't attention-churn the rest of the team, that would be much appreciated. 🙂

Sorry about that, I was trying to describe what I thought before open direct PR. it seems I didn't do well.

> ...what I _would_ like to see is a proposal that _demonstrates_ that change in a minimal way.

Yes, I'll try to do it.

**Comment by lovelydinosaur** [comment, marked as answer]
I think the most graceful way to do what you're attempting here would be adding the following in `_syncronization.py`...

```python
EXCEPTION_OR_CANCELLED = (Exception,)

try:
    import trio  # trio support is available to httpcore
    EXCEPTION_OR_CANCELLED += (trio.CancelledError,)
except ImportError:
    pass

try:
    import anyio  # asyncio support is available to httpcore
    import asyncio
    EXCEPTION_OR_CANCELLED += (asyncio.Cancelled,)
except ImportError:
    pass
```

And then import and use EXCEPTION_OR_CANCELLED in the existing `BaseException` cases...

```python
try:
    ...
except EXCEPTION_OR_CANCELLED as exc:
    ...
```

The would seem to me to be the simplest approach that would work fine for both the `_sync` and `_async` codebases.

**Reply by T-256** [reply]
Thanks
